gematria/datasets/basic_block_utils.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/basic_block_utils.h"
+
+#include <map>
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "llvm/MC/MCInstrInfo.h"
+#include "llvm/MC/MCRegisterInfo.h"
+
+using namespace llvm;
+
+namespace gematria {
+
+unsigned getSuperRegister(unsigned OriginalRegister,
+                          const MCRegisterInfo &RegisterInfo) {
+  unsigned SuperRegister = OriginalRegister;
+  for (MCPhysReg CurrentSuperRegister :
+       RegisterInfo.superregs_inclusive(OriginalRegister)) {
+    SuperRegister = CurrentSuperRegister;
+  }
+  // Only return super registers for GPRs. Since mainly GPRs will be used for
+  // addressing, redefining other aliasing registers (like vector registers)
+  // does not matter as much.
+  if (RegisterInfo.getRegClass(X86::GR64RegClassID).contains(SuperRegister))
+    return SuperRegister;
+  else
+    return OriginalRegister;
+}
+
+std::vector<unsigned> BasicBlockUtils::getUsedRegisters(
+    const std::vector<DisassembledInstruction> &Instructions,
+    const MCRegisterInfo &RegisterInfo, const MCInstrInfo &InstructionInfo) {
+  std::map<unsigned, bool> UsedInputRegisters;
+  std::map<unsigned, bool> DefinedInBBRegisters;
```

Comment:
These two are defined as maps, but they are used as sets. It would be more efficient to use `std::set`.

But given that the number of registers is fairly small, it might be even faster to use `llvm::BitVector` that has a very nice cache profile and very fast access (ideally, `std::vector<bool>` but that doesn't have a good API for iterating over all ones).

Reply:
Switch to `std::set` for now.